### PR TITLE
imlement FST support for verilator

### LINKED ISSUE #2131: Simulation fail when enabling Verilator FST trace
Verilator supports FST format tracing through the --trace-fst option.  However, when this option is added, cocotb fails to build it with the following error:

    /usr/lib/python3.8/site-packages/cocotb/share/lib/verilator/verilator.cpp: In function ‘int main(int, char**)’:
    /usr/lib/python3.8/site-packages/cocotb/share/lib/verilator/verilator.cpp:41:23: error: cannot convert ‘std::unique_ptr<VerilatedVcdC>::pointer’ {aka ‘VerilatedVcdC*’} to ‘VerilatedFstC*’
       41 |     top->trace(tfp.get(), 99);
          |                ~~~~~~~^~
          |                       |
          |                       std::unique_ptr<VerilatedVcdC>::pointer {aka VerilatedVcdC*}
    In file included from /usr/lib/python3.8/site-packages/cocotb/share/lib/verilator/verilator.cpp:5:
    ./Vtop.h:235:31: note:   initializing argument 1 of ‘void Vtop::trace(VerilatedFstC*, int, int)’
      235 |     void trace(VerilatedFstC* tfp, int levels, int options = 0);
          |                ~~~~~~~~~~~~~~~^~~
    make[2]: *** [Vtop.mk:61: verilator.o] Error 1

Using --trace builds fine, but the resulting VCD traces are many GB in size.

### PR DESCRIPTION
**Pull request by markusdd**
This addresses Ticket #1709 

The VM_TRACE define is set by verilator and its value cannot be modified.
Cleanest solution seems to be to add another orthogonal define via CFLAGS.

Documentation is also adapted, because verilator also needs to be called via --trace-fst instead of --trace

I have checked their docs and there is no seperate define being set when FST option is selected, so
we have to do it by ourselves for now.

Closes #2131.

### CODE REVIEW

**Review by ktbarrett — COMMENTED** [review]
It's kind of a shame you have to do this. It seems like verilator could offer a generic facade and make the type of tracer testable with VM_TRACE.

**Comment by markusdd** [comment]
> It's kind of a shame you have to do this. It seems like verilator could offer a generic facade and make the type of tracer testable with VM_TRACE.

True, although I know that FST can do more than VDC, so the extra class might be warranted to provide a richer interface.

But the define thing definetly should be addressed.

Should I create an upstream issue?

**Comment by ktbarrett** [comment]
@markusdd since you have been working on this, can I ask if trace VCD functionality actually works for you? I can't seem to get it to actual dump anything...

**Comment by markusdd** [comment]
You mean based on this PR or generally?

Have you added the --trace --trace-structs options?

(I use the Makefile style like in the debug repo I made for the other issue, with that it works for me)

**Comment by ktbarrett** [comment]
@markusdd Generally. Running in your repo with a test that just runs the clock, like I have below, results in [dump.vcd.txt](https://github.com/cocotb/cocotb/files/5341165/dump.vcd.txt). Nothing is actually dumped, despite the signal actually changing (otherwise the test wouldn't run to completion). I am using Verilator master with my patches for the bug he have previously discussed. I will try with this branch dumping in FST and see if I get different results tonight.

```python
@cocotb.test()
async def test1(dut):

    async def clock():
        while True:
            dut.clk_w_i <= 1
            await Timer(10, 'ns')
            dut.clk_w_i <= 0
            await Timer(10, 'ns')
    cocotb.fork(clock())

    for i in range(100):
        await RisingEdge(dut.clk_w_i)
```

**Comment by markusdd** [comment]
what does the verilated verilog look like?

Maybe you are the victim of an optimization if there is nothing in the module that would warrant keeping it.

**Comment by markusdd** [comment]
Your test yields this in my environment, so it works:
grafik

**Comment by ktbarrett** [comment]
Fun... is it possible you can write a self checking test for this feature? Since I won't be able to replicate it myself until I track that down.

**Comment by markusdd** [comment]
Question is what such a test would look like.
It is hard to check the validity of a waveform dump without looking at it...

**Comment by imphil** [comment]
> It is hard to check the validity of a waveform dump without looking at it...

There's vcddiff (https://github.com/veripool/vcddiff), used in Verilator's test suite, for example.

**Comment by markusdd** [comment]
That's an interesting find.

How would this integrate with your test flow/env setup.

I'm not familiar with this, the test would probably need to be running a cocotb test wrappen in a script and then compare against an existing vcd (also: wehere would we store the vcd, we don't want this in the repo I guess, those guys are huge)

**Comment by markusdd** [comment]
@ktbarrett what is your take here, how should we setup the testing.
(I do not wanna do something which doesn't fit the scheme and throw it away)

I see 4 issues:

1) How to integrate that test into the flow
2) where to store a sample vcd, these are huge
3) shoudl we rely on something external like vcd-diff
4) does a vcd test belong into this PR?

**Comment by markusdd** [comment]
Ok, so good news, we now have upstream support to get the VM_TRACE_FST define, my merge request was accepted: https://github.com/verilator/verilator/pull/2592

So using the -CFLAGS way remains optional, but is a good way for users who are forced to use a verilator version that is a bit older.

**Comment by ktbarrett** [comment]
@markusdd I'm reconsidering doing that. It is a lot of work and will require us to depend upon non-essential libraries to do regression testing. Since this only affects one simulator doing one type of waveform tracing, I think it's sufficient to just upload pictures of it working on your end. So nevermind 😄.

**Comment by markusdd** [comment]
That is where I'm at as well.
This will be a lot of effort for little gain overall.

I can get you an FST, still I would like to understand why this doesn't work on your end.

**Comment by ktbarrett** [comment]
I am building verilator from source. Is there a particular directive I need to give to `configure` to enable tracing support?

**Comment by markusdd** [comment]
[dump.fst.zip](https://github.com/cocotb/cocotb/files/5373058/dump.fst.zip)

So here is an fst. (had to zip it for guthub to accept it)

**Comment by markusdd** [comment]
> I am building verilator from source. Is there a particular directive I need to give to `configure` to enable tracing support?

Can you check your configure output? It will skip vcd and especially fst tracing support when there are libraries missing.

**Comment by ktbarrett** [comment]
Ubuntu 20.04 install on WSL, I built master as of 1 week ago. I'll try it on Debian unstable later, building verilator takes a while on my old T420...

**Comment by markusdd** [comment]
> Ubuntu 20.04 install on WSL, I built master as of 1 week ago. I'll try it on Debian unstable later, building verilator takes a while on my old T420...

OK WSL might have a few quirks there, that is possible.
I've seen it working under MacOS and Fedora 32.

File: documentation/source/simulator_support.rst
Comment (themperek) [inline documentation/source/simulator_support.rst:128]: Since https://github.com/verilator/verilator/pull/2592 will it work like this?
```suggestion
    EXTRA_ARGS += --trace-fst --trace-structs
```

File: cocotb/share/lib/verilator/verilator.cpp
Comment (imphil) [inline cocotb/share/lib/verilator/verilator.cpp:11]: Please try to match the new Verilator behavior here, which unconditionally defines VM_TRACE_FST and sets it to 1 or 0. For backwards-compat, introduce a block like this:

```c
#ifndef VM_TRACE_FST
    #define VM_TRACE_FST 0
#endif
```

**Review by imphil — COMMENTED** [review]
Thanks for your contribution @markusdd. LGTM in general with a couple of small comments. Please have a look at those. Can you also combine all your commits into a single one with a descriptive commit message (see https://github.com/cocotb/cocotb/blob/master/CONTRIBUTING.md#patch-requirements). Please let us know if you're getting lost in git commands, and I'll have a look.

File: documentation/source/simulator_support.rst
Comment (themperek) [reply documentation/source/simulator_support.rst:128]: Although at the moment of cocotb realase it may be ok.

Comment (markusdd) [reply documentation/source/simulator_support.rst:128]: We should keep the -CFLAGS hint.
The newest verilator will not be in major distributions for at least a couple of months, I would at least make mention of this possibility.

File: documentation/source/simulator_support.rst
Comment (markusdd) [reply documentation/source/simulator_support.rst:135]: Yes but in fact does not apply to vcd. vcd is fine on older gtkwave and verilator.
The only real improvements happendes very recently in verialtor and gtkwave (actually gtkwave is more important in that context as it provides the fst library)

File: cocotb/share/lib/verilator/verilator.cpp
Comment (markusdd) [reply cocotb/share/lib/verilator/verilator.cpp:11]: does that do anything useful in that context? 

Shouldn't make a difference even when using #if vs. #ifdef?

File: documentation/source/simulator_support.rst
Comment (markusdd) [reply documentation/source/simulator_support.rst:132]: see below.

Yes it does, but I would not move to that just yet, we need to mention you can 'workaround' a slightly older verilator version like this.

File: cocotb/share/lib/verilator/verilator.cpp
Comment (ktbarrett) [reply cocotb/share/lib/verilator/verilator.cpp:11]: When you enable FST tracing is `VM_TRACE` defined? I think he is saying you can lift the `if VM_TRACE_FST` out of this `if`.

```c++
#if VM_TRACE_FST
#include <verilated_fst_c.h>
#elif VM_TRACE
#include <verilated_vcd_c.h>
#endif
```

File: cocotb/share/lib/verilator/verilator.cpp
Comment (ktbarrett) [inline cocotb/share/lib/verilator/verilator.cpp:51]: And you could do the same here.

File: cocotb/share/lib/verilator/verilator.cpp
Comment (markusdd) [reply cocotb/share/lib/verilator/verilator.cpp:11]: Hi, to quote myself:

> The VM_TRACE define is set by verilator and its value cannot be modified.
Cleanest solution seems to be to add another orthogonal define via CFLAGS.

That was the whole point of using an extra define because VM_TRACE always appears.

File: cocotb/share/lib/verilator/verilator.cpp
Comment (markusdd) [reply cocotb/share/lib/verilator/verilator.cpp:51]: see above

File: cocotb/share/lib/verilator/verilator.cpp
Comment (ktbarrett) [reply cocotb/share/lib/verilator/verilator.cpp:11]: I'm not sure what the problem is with this code. It looks fine to me. I'm partial to merge this anyways unless @imphil explain a bit better. I'll give it 2 days. This is certainly not the last time we will have to deal with Verilator support, it doesn't have to be perfect now, but hopefully will be before 1.5 release at the end of the year.

Comment (imphil) [reply cocotb/share/lib/verilator/verilator.cpp:11]: > does that do anything useful in that context?

Yes. The goal is to get exactly the same defines for older and newer Verilator versions.

* For newer Verilator versions, there's unconditionally a define VM_TRACE_FST set to either 0 or 1. 
* For older Verilator versions, you recommend in the docs to set `-DVM_TRACE_FST=1` iff FST tracing is wanted. This means, VM_TRACE_FST is only set to 1 if FST tracing is enabled, and undefined otherwise.

If you add the code snippet I've shown, you get identical behavior.

Now I know that this doesn't make much of a difference given the way the pre-processor treats undefined macros, but I feel it's better to be explicit here.

File: documentation/source/simulator_support.rst
Comment (imphil) [inline documentation/source/simulator_support.rst:132]:
```suggestion
    EXTRA_ARGS += --trace-fst --trace-structs -CFLAGS -DVM_TRACE_FST=1
```

to match your advice above.

File: documentation/source/simulator_support.rst
Comment (imphil) [inline documentation/source/simulator_support.rst:135]: GtkWave has been supporting FST for ages (ever?), and Verilator supported it also for multiple years (May 2018 actually, see https://github.com/verilator/verilator-announce/issues/17). Just remove this sentence. Earlier versions don't work with cocotb anyways.

Comment (markusdd) [reply documentation/source/simulator_support.rst:135]: That is not completely accurate as I've experienced myself. There have been relatively recent updates (meaning this year) which have fixed a few issues that allowed it to be used. Before that performance was abysmal.

But I do not care, we can remove it, I just thought it might be a helpful hint.

File: documentation/source/simulator_support.rst
Comment (markusdd) [reply documentation/source/simulator_support.rst:132]: Fine by me.

**Review by imphil — APPROVED** [review]
LGTM, please squash into a single commit with a descriptive commit message before merging.

**Comment by markusdd** [comment]
> 
> 
> LGTM, please squash into a single commit with a descriptive commit message before merging.

will do tonight, working on the wrong machine currently.

**Comment by ktbarrett** [comment]
@markusdd Squashing it something I can do when I merge the PR, I did it on your last one too =)

**Comment by markusdd** [comment]
> 
> 
> @markusdd Squashing it something I can do when I merge the PR, I did it on your last one too =)

even better. Just out of curiosity, can this be done online here in Github? (I'm more of a Gitlab user in my daily job, so I do not know all ins and outs of this here.)

**Comment by ktbarrett** [comment]
I don't think you can do that from the Github UI. Maintainers, when accepting a PR, get to add the change to the commit tree by either merging (creating a merge commit), rebasing all of the commits on master, or squashing all the commits into 1 and rebasing that on master. People can do it manually, but it's equally easy for the Maintainer to do it from the UI. Sometimes they do it because there was a lot of "applied suggestions from code review" commits; or they want to modify/improve a commit message.

**Comment by markusdd** [comment]
> 
> 
> I don't think you can do that from the Github UI. Maintainers, when accepting a PR, get to add the change to the commit tree by either merging (creating a merge commit), rebasing all of the commits on master, or squashing all the commits into 1 and rebasing that on master. People can do it manually, but it's equally easy for the Maintainer to do it from the UI. Sometimes they do it because there was a lot of "applied suggestions from code review" commits; or they want to modify/improve a commit message.

Makes sense, thanks for the explanation.

**Pull request merged**
